File: swiftui/CellposeGUI/Models/GuiSyncScope.swift
import Foundation

struct GuiSyncScope: OptionSet {
    let rawValue: Int

    static let labels = GuiSyncScope(rawValue: 1 << 0)
    static let canvasImage = GuiSyncScope(rawValue: 1 << 1)
    static let canvasMask = GuiSyncScope(rawValue: 1 << 2)
    static let canvasSelection = GuiSyncScope(rawValue: 1 << 3)

    static let selectionChange: GuiSyncScope = [.canvasSelection]
    static let instanceEdit: GuiSyncScope = [.labels, .canvasMask, .canvasSelection]
    static let filterChange: GuiSyncScope = instanceEdit
    static let visibilityEdit: GuiSyncScope = [.labels, .canvasMask]
    static let maskGeometry: GuiSyncScope = [.labels, .canvasMask]
}

File: winui/CellposeGUI/Models/GuiSyncScope.cs
namespace CellposeGUI.Models;

[Flags]
public enum GuiSyncScope
{
    None = 0,
    Labels = 1 << 0,
    VisibilityHeader = 1 << 1,
    CanvasImage = 1 << 2,
    CanvasMask = 1 << 3,
    CanvasSelection = 1 << 4,
    Ellipses = 1 << 5,

    SelectionChange = CanvasSelection,
    InstanceEdit = Labels | CanvasMask | CanvasSelection,
    FilterChange = InstanceEdit,
    VisibilityEdit = Labels | CanvasMask | VisibilityHeader,
    MaskGeometry = Labels | CanvasMask | Ellipses,
}
